doc/sshguard.8
.\" Man page generated from reStructuredText.
.
.TH SSHGUARD 8 "August 16, 2016" "2.0.0" "SSHGuard Manual"
.SH NAME
sshguard \- block brute-force attacks by aggregating system logs
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.\" Copyright (c) 2007,2008,2009,2010 Mij <mij@sshguard.net>
.
.\" Permission to use, copy, modify, and distribute this software for any
.
.\" purpose with or without fee is hereby granted, provided that the above
.
.\" copyright notice and this permission notice appear in all copies.
.
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.
.SH SYNOPSIS
.sp
\fBsshguard\fP [\fB\-v\fP] [\fB\-h\fP]
[\fB\-a\fP \fIblacklist\-threshold\fP]
[\fB\-b\fP \fIblacklist\-file\fP]
[\fB\-i\fP \fIpid\-file\fP]
[\fB\-p\fP \fIblock\-time\fP]
[\fB\-s\fP \fIdetection\-time\fP]
[\fB\-w\fP \fIaddress\fP | \fIfile\fP]
[\fIfile\fP ...]
.SH DESCRIPTION
.sp
\fBsshguard\fP protects hosts from brute\-force attacks against SSH and other
services. It aggregates system logs and blocks repeat offenders using one of
several firewall backends, including \fBfirewalld\fP, \fBipfw\fP, \fBipset\fP,
\fBiptables\fP and \fBpf\fP\&.
.sp
\fBsshguard\fP can read log messages from standard input (suitable for piping
from \fBsyslog\fP or \fBjournalctl\fP) or monitor one or more log files. Log
messages are parsed, line\-by\-line, for recognized patterns. If an attack,
such as several login failures within a few seconds, is detected, the
offending IP is blocked. Offenders are unblocked after a set interval, but
can be semi\-permanently banned using the blacklist option.
.sp
See \fI\%http://www.sshguard.net/docs/setup/\fP for setup instructions.
.sp
Other features, attack signatures, and additional documentation can be found
at \fI\%http://www.sshguard.net/\fP\&.
.SH OPTIONS
.INDENT 0.0
.TP
.B \fB\-a\fP \fIblacklist\-threshold\fP (default 30)
Block an attacker when its dangerousness exceeds \fIblacklist\-threshold\fP\&.
Each attack pattern that is matched contributes a fixed dangerousness
of 10.
.TP
.B \fB\-b\fP \fIblacklist\-file\fP
Blacklist an attacker when its dangerousness exceeds \fIblacklist\-threshold\fP\&.
Blacklisted addresses are added to \fIblacklist\-file\fP so they can be read at
the next startup. Blacklisted addresses are never automatically unblocked,
but it is good practice to periodically clean out stale blacklist entries.
.TP
.B \fB\-i\fP \fIpid\-file\fP
Write the PID of \fBsshguard\fP to \fIpidfile\fP\&.
.TP
.B \fB\-p\fP \fIblock\-time\fP (default 120 secs, or 2 minutes)
Wait at least \fIblock\-time\fP seconds before releasing a blocked address.
Repeat attackers are blocked for 1.5 times longer after each attack.
Because \fBsshguard\fP unblocks attackers only at infrequent intervals,
this parameter is inexact (actual blocks will be longer).
.TP
.B \fB\-s\fP \fIdetection\-time\fP (default 1800 secs, or 30 minutes)
Forget about an attacker \fIdetection\-time\fP seconds after its last attempt.
Its  dangerousness will be reset to zero.
.TP
.B \fB\-w\fP \fIip\-address\fP | \fIwhitelist\-file\fP
Whitelist the given address, hostname, or address block. Alternatively,
read whitelist entires from \fIwhitelist\-file\fP\&. This option can be given
multiple times. See WHITELISTING below for details.
.TP
.B \fB\-h\fP
Print usage information and exit.
.TP
.B \fB\-v\fP
Print version information and exit.
.UNINDENT
.SH ENVIRONMENT
.INDENT 0.0
.TP
.B SSHGUARD_DEBUG
Enable additional debugging information.
.UNINDENT
.SH WHITELISTING
.sp
\fBsshguard\fP supports IP address whitelisting. Whitelisted addresses are not
blocked even if they appear to generate attacks. This is useful for protecting
lame LAN users (or external friendly users) from being incidentally blocked.
.sp
Whitelist addresses are controlled through the \-w command\-line option. This
option can add explicit addresses, host names and address blocks:
.INDENT 0.0
.TP
.B addresses
specify the numeric IPv4 or IPv6 address directly, like:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
\-w 192.168.1.10
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
or in multiple occurrences:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
\-w 192.168.1.10 \-w 2001:0db8:85a3:0000:0000:8a2e:0370:7334
.ft P
.fi
.UNINDENT
.UNINDENT
.TP
.B host names
specify the host name directly, like:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
\-w friendhost.enterprise.com
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
or in multiple occurrences:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
\-w friendhost.enterprise.com \-w friend2.enterprise.com
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
All IPv4 and IPv6 addresses that the host resolves to are whitelisted. Hosts
are resolved to addresses once, when \fBsshguard\fP starts up.
.TP
.B address blocks
specify the IPv4 or IPv6 address block in the usual CIDR notation:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
\-w 2002:836b:4179::836b:0000/126
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
or in multiple occurrences:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
\-w 192.168.0.0/24 \-w 1.2.3.128/26
.ft P
.fi
.UNINDENT
.UNINDENT
.TP
.B file
When longer lists are needed for whitelisting, they can be wrapped into a
plain text file, one address/hostname/block per line, with the same syntax
given above.
.sp
\fBsshguard\fP can take whitelists from files when the \-w option argument begins
with a \(aq.\(aq (dot) or \(aq/\(aq (slash).
.sp
This is a sample whitelist file (say /etc/friends):
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
# comment line (a \(aq#\(aq as very first character)
#   a single IPv4 and IPv6 address
1.2.3.4
2001:0db8:85a3:08d3:1319:8a2e:0370:7344
#   address blocks in CIDR notation
127.0.0.0/8
10.11.128.0/17
192.168.0.0/24
2002:836b:4179::836b:0000/126
#   hostnames
rome\-fw.enterprise.com
hosts.friends.com
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
And this is how \fBsshguard\fP is told to make a whitelist up from the
/etc/friends file:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
sshguard \-w /etc/friends
.ft P
.fi
.UNINDENT
.UNINDENT
.UNINDENT
.sp
The \-w option can be used only once for files. For addresses, host names and
address blocks it can be used with any multiplicity, even with mixes of them.
.SH SEE ALSO
.sp
Glossary: \fI\%http://www.sshguard.net/docs/terminology/\fP
.sp
Website: \fI\%http://www.sshguard.net/\fP
.sp
\fBsshguard.conf.sample\fP
.SH AUTHORS
.sp
Michele Mazzucchi <\fI\%mij@bitchx.it\fP>,
T.J. Jones <\fI\%tjjones03@gmail.com\fP>,
Kevin Zheng <\fI\%kevinz5000@gmail.com\fP>
.\" Generated by docutils manpage writer.
.


Changes
* 1.6
    - blacklist: blacklist file is now human-readable/editable (thanks Armando Miraglia)
    - fix Log Sucker bug with filesystems with gazillion files (thanks Johann H. Hauschild)
    - fix syslog format detection on hosts with undefined hostname
    - fix ipfw backend crashing when blacklist is enabled but empty on start-up (thanks Jin Choi)
    - hosts backend: ensure safe permission for tcpwrappers file regardless of the local umask

* 1.5       Feb 2011
    - logsucker: sshguard polls multiple log files at once
    - recognize syslog's "last message repeated N times" contextually and per-source
    - attackers now gauged with attack *dangerousness* instead of count (adjust your -a !)
    - improve IPv6 support
    - add detection for: Exim, vsftpd, Sendmail, Cucipop
    - improve Solaris support (thanks OpenCSW.org folks)
    - handle huge blacklists efficiently
    - improve logging granularity and descriptiveness
    - add -i command line option for saving PID file as an aid for startup scripts
    - update some attack signatures
    - many other improvements, see 1.5beta and 1.5rc changelogs for complete credits
    - fix a recognition problem for multilog files
    - fix log filtering on OSes with inverted priority declarations
    - fix file descriptor leak if "ps" command fails to run
    - fix whitelist module allowing some entries to be skipped (thanks Andrea Dal Farra)
    - fix segfault from invalid free() when all DNS lookups fail
    - fix assertion failure when logsucker is notified before the logging completes (thanks Colin Keith)

* 1.4       Aug 2009
    - add touchiness: block repeated abusers for longer
    - add blacklisting: store frequent abusers for permanent blocking
    - add support for IPv6 in whitelisting (experimental)
    - sshguard ignores interrupted fgets() and reloads more seldom (thanks Keven Tipping)
    - debug mode now enabled with SSHGUARD_DEBUG environment variable (no "-d")
    - support non-POSIX libCs that require getopt.h (thanks Nobuhiro Iwamatsu)
    - import newer SimCList containing a number of fixes and improvements
    - firewall backends now block all traffic from attackers by default, not per-service
    - netfilter/iptables backend now verifies credentials at initialization
    - parser accepts "-" and "_" chars in process names
    - fix detection of some ProFTPd and pure-ftp messages
    - support log formats of new versions of ProFTPd
    - fix one dovecot pattern
    - correctly handle abuse threshold = 1 (thanks K. Tipping)
    - fix handling of IPv6 with IPFW under Mac OS X Leopard (thanks David Horn)
    - fix cmdline argument BoF exploitable by local users when sshguard is setuid
    - support blocking IPv6 addrs in backed "hosts.allow"
    - extend hosts.allow backend to support all service types
    - localhost addresses are now whitelisted a priori
    - extend IPv6 pattern for matching special addresses (eg, IPv4 embedded)
    - fix grammar to be insensitive to a log injection in sshd (thanks J. Oosterveen)

* 1.3       Oct 2008
    - fix autoconf problem
    - automatically detect when ipfw supports IPv6 (thanks David Horn)
    - be sensitive to proftpd messages to auth facility, not daemon (thanks Andy Berkvam)
    - add sshd pattern for "Bad protocol" and "Did not receive identif string"

* 1.2       Sep 2008
    - support for Cyrus IMAP
    - support for SSH "possible break-in attempt" messages
    - updated support for dovecot to include logging format of new versions
    - (thanks Michael Maynard) fix of IPF backend causing sshguard not to
          update /etc/ipf.rules (disallow IPv6)
    - fix detection of password when sshd doesn't log anything more than PAM

* 1.1       Jul 2008 (midway releases from Jul 2007 to Jun 2008)
    - support suspension
    - support debug mode at runtime (-d) for helping users in problem solving
    - support for metalog logging format
    - fix parser bug when recognizing certain IPv6 addresses
    - fix segfault when the pipe to sshguard is closed unexpectedly
    - support for ipfilter as blocking backend (thanks Hellmuth Michaelis for feedback)
    - support for log messages authentication
    - support for AIX genfilt firewall (thanks Gabor Szittner)
    - fix "hosts" backend bug not discarding temporary files
    - add monitoring support for new services:
        @ dovecot imap
        @ UWimap imap and pop
        @ FreeBSD's ftpd
        @ ProFTPd
        @ pure-ftpd

* 1.0       May 2007
    - address whitelisting for protecting friend addressess
    - support for IPv6
    - support for service multiplexing (behave differently for different services)
    - more powerful parsing (context-free): support multilog, autotranslate
        hostnames and easily extends to a lot of services
    - new blocking backend: "hosts" for /etc/hosts.deny
    - paths autodetected and adjustable from ./configure
    - script for trivially generating new custom backends

* 0.91      Mar 2007
    - run away from scons and use autotools as building system

* 0.9       Feb 2007
    - first public release



src/blocker/sshguard.c
/*
 * Copyright (c) 2007,2008,2009,2010 Mij <mij@sshguard.net>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

/*
 * SSHGuard. See http://www.sshguard.net
 */

#include "config.h"

#include <assert.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "blocklist.h"
#include "fw.h"
#include "parser/parser.h"
#include "sandbox.h"
#include "simclist.h"
#include "sshguard.h"
#include "sshguard_blacklist.h"
#include "sshguard_log.h"
#include "sshguard_options.h"
#include "sshguard_whitelist.h"

/** Keep track of the exit signal received. */
static volatile sig_atomic_t exit_sig = 0;

/*      FUNDAMENTAL DATA STRUCTURES         */
/* These lists are all lists of attacker_t structures.
 * limbo and hell maintain "temporary" entries: in limbo, entries are deleted
 * when the address is detected to have abused a service (right after it is
 * blocked); in hell, it is deleted when the address is released.
 *
 * The list offenders maintains a permanent history of the abuses of
 * attackers, their first and last attempt, the number of abuses etc. These
 * are maintained for entire runtime. When the number of abuses exceeds a
 * limit, an address might be blacklisted (if blacklisting is enabled with
 * -b). After blacklisting, the block of an attacker is released, because it
 *  has already been blocked permanently.
 *
 *  The invariant of "offenders" is: it is sorted in decreasing order of the
 *  "whenlast" field.
 */
/* list of addresses that failed some times, but not enough to get blocked */
list_t limbo;
/* list of offenders (addresses already blocked in the past) */
list_t offenders;

/* handler for termination-related signals */
static void sigfin_handler();
/* called at exit(): flush blocked addresses and finalize subsystems */
static void finishup(void);

/* handle an attack: addr is the author, addrkind its address kind, service the attacked service code */
static void report_address(attack_t attack);
/* cleanup false-alarm attackers from limbo list (ones with too few attacks in too much time) */
static void purge_limbo_stale(void);

static void my_pidfile_create() {
    FILE *p = fopen(opts.my_pidfile, "w");
    if (p == NULL) {
        sshguard_log(LOG_ERR, "Failed to create pid file: %m");
        exit(73);
    }

    fprintf(p, "%d\n", (int)getpid());
    fclose(p);
}

static void my_pidfile_destroy() {
    if (unlink(opts.my_pidfile) != 0) {
        sshguard_log(LOG_ERR, "Failed to remove pid file: %m");
    }
}

int main(int argc, char *argv[]) {
    int sshg_debugging = (getenv("SSHGUARD_DEBUG") != NULL);
    sshguard_log_init(sshg_debugging);

    srand(time(NULL));

    /* pending, blocked, and offender address lists */
    list_init(&limbo);
    list_attributes_seeker(& limbo, attack_addr_seeker);
    blocklist_init();
    list_init(&offenders);
    list_attributes_seeker(& offenders, attack_addr_seeker);
    list_attributes_comparator(& offenders, attackt_whenlast_comparator);

    // Initialize whitelist before parsing arguments.
    whitelist_init();

    if (get_options_cmdline(argc, argv) != 0) {
        exit(64);
    }

    if (opts.my_pidfile != NULL) {
        my_pidfile_create();
        atexit(my_pidfile_destroy);
    }

    if (fw_init() != FWALL_OK) {
        sshguard_log(LOG_ERR, "Failed to initialize firewall");
        exit(69);
    }

    if (opts.blacklist_filename != NULL) {
        blacklist_load_and_block();
    }

    /* termination signals */
    signal(SIGTERM, sigfin_handler);
    signal(SIGHUP, sigfin_handler);
    signal(SIGINT, sigfin_handler);
    atexit(finishup);

    sandbox_init();

    /* whitelist localhost */
    if (whitelist_add("127.0.0.1") != 0) {
        fprintf(stderr, "Could not whitelist localhost. Terminating...\n");
        exit(1);
    }

    whitelist_conf_fin();

    sshguard_log(LOG_INFO, "Monitoring attacks");

    char buf[1024];
    attack_t parsed_attack;
    while (fgets(buf, sizeof(buf), stdin) != NULL) {
        if (sscanf(buf, "%d %46s %d %d\n", &parsed_attack.service,
                  parsed_attack.address.value, &parsed_attack.address.kind,
                  &parsed_attack.dangerousness) == 4) {
            sshguard_log(LOG_DEBUG,
                         "Attack from %s on service %d with danger %u",
                         parsed_attack.address.value, parsed_attack.service,
                         parsed_attack.dangerousness);
            report_address(parsed_attack);
        } else {
            sshguard_log(LOG_ERR, "Could not parse attack data");
            exit(65);
        }
    }

    if (feof(stdin)) {
        sshguard_log(LOG_NOTICE, "Received EOF from stdin");
    }
}

void log_block(attacker_t *tmpent, attacker_t *offenderent) {
    char time_msg[128] = "forever";
    const time_t time = tmpent->pardontime;
    if (time > 0) {
        if (snprintf(time_msg, sizeof(time_msg), "for %lld secs", (long long)time) < 0) {
            abort();
        }
    }
    sshguard_log(LOG_NOTICE, "Blocking %s %s (%u attacks in %lld "
                             "secs, after %d abuses over %lld secs)",
                 tmpent->attack.address.value, time_msg, tmpent->numhits,
                 (long long)(tmpent->whenlast - tmpent->whenfirst),
                 offenderent->numhits,
                 (long long)(offenderent->whenlast - offenderent->whenfirst));
}

/*
 * This function is called every time an attack pattern is matched.
 * It does the following:
 * 1) update the attacker infos (counter, timestamps etc)
 *      --OR-- create them if first sight.
 * 2) block the attacker, if attacks > threshold (abuse)
 * 3) blacklist the address, if the number of abuses is excessive
 */
static void report_address(attack_t attack) {
    attacker_t *tmpent = NULL;
    attacker_t *offenderent;

    assert(attack.address.value != NULL);
    assert(memchr(attack.address.value, '\0', sizeof(attack.address.value)) != NULL);

    /* clean list from stale entries */
    purge_limbo_stale();

    /* address already blocked? (can happen for 100 reasons) */
    if (blocklist_contains(attack)) {
        sshguard_log(LOG_WARNING, "%s has already been blocked",
                attack.address.value);
        return;
    }

    if (whitelist_match(attack.address.value, attack.address.kind)) {
        sshguard_log(LOG_INFO, "%s: not blocking (on whitelist)",
                attack.address.value);
        return;
    }
    
    /* search entry in list */
    tmpent = list_seek(& limbo, & attack.address);
    if (tmpent == NULL) { /* entry not already in list, add it */
        /* otherwise: insert the new item */
        tmpent = malloc(sizeof(attacker_t));
        attackerinit(tmpent, & attack);
        list_append(&limbo, tmpent);
    } else {
        /* otherwise, the entry was already existing, update with new data */
        tmpent->whenlast = time(NULL);
        tmpent->numhits++;
        tmpent->cumulated_danger += attack.dangerousness;
    }

    if (tmpent->cumulated_danger < opts.abuse_threshold) {
        /* do nothing now, just keep an eye on this guy */
        return;
    }

    /* otherwise, we have to block it */
    

    /* find out if this is a recidivous offender to determine the
     * duration of blocking */
    tmpent->pardontime = opts.pardon_threshold;
    offenderent = list_seek(& offenders, & attack.address);
    if (offenderent == NULL) {
        /* first time we block this guy */
        sshguard_log(LOG_DEBUG, "%s: first block (adding as offender)",
                tmpent->attack.address.value);
        offenderent = (attacker_t *)malloc(sizeof(attacker_t));
        /* copy everything from tmpent */
        memcpy(offenderent, tmpent, sizeof(attacker_t));
        /* adjust number of hits */
        offenderent->numhits = 1;
        list_prepend(& offenders, offenderent);
        assert(! list_empty(& offenders));
    } else {
        /* this is a previous offender, update dangerousness and last-hit timestamp */
        offenderent->numhits++;
        offenderent->cumulated_danger += tmpent->cumulated_danger;
        offenderent->whenlast = tmpent->whenlast;
    }

    /* At this stage, the guy (in tmpent) is offender, and we'll block it anyway. */

    /* Let's see if we _also_ need to blacklist it. */
    if (opts.blacklist_filename != NULL &&
        offenderent->cumulated_danger >= opts.blacklist_threshold) {
        /* this host must be blacklisted -- blocked and never unblocked */
        tmpent->pardontime = 0;

        /* insert in the blacklisted db iff enabled */
        if (opts.blacklist_filename != NULL) {
            blacklist_add(offenderent);
        }
    } else {
        /* compute blocking time wrt the "offensiveness" */
        for (unsigned int i = 0; i < offenderent->numhits; i++) {
            tmpent->pardontime *= 2;
        }
    }
    list_sort(& offenders, -1);
    log_block(tmpent, offenderent);

    /* append blocked attacker to the blocked list, and remove it from the pending list */
    blocklist_add(tmpent);
    assert(list_locate(& limbo, tmpent) >= 0);
    list_delete_at(& limbo, list_locate(& limbo, tmpent));
}

static void purge_limbo_stale(void) {
    sshguard_log(LOG_DEBUG, "Purging old attackers");
    time_t now = time(NULL);
    for (unsigned int pos = 0; pos < list_size(&limbo); pos++) {
        attacker_t *tmpent = list_get_at(&limbo, pos);
        if (now - tmpent->whenfirst > opts.stale_threshold) {
            list_delete_at(&limbo, pos);
            free(tmpent);
            pos--;
        }
    }
}

static void finishup(void) {
    sshguard_log(LOG_INFO, "Exiting on %s",
            exit_sig == SIGHUP ? "SIGHUP" : "signal");
    fw_fin();
    whitelist_fin();
    sshguard_log_fin();
}

static void sigfin_handler(int sig) {
    exit_sig = sig;
    exit(0);
}
